File: lang/src/main/java/uk/co/zentium/langjava/Succinct1.java
package uk.co.zentium.langjava;

import java.util.*;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class Succinct1 {

    private static Map<String, List<String>> validKeyMap = new HashMap<>();

    static {
        validKeyMap.put("logo", Arrays.asList("src"));
        validKeyMap.put("uuid",  Arrays.asList("cgid", "src"));
        validKeyMap.put("pcms", Arrays.asList("cgid", "src"));
    }

    private static Boolean hasValidAttributeKeys(String idType, Map<String, String> attributes) {

        return attributes.keySet()
                .stream()
                .mapToLong(item ->
                        validKeyMap.keySet().contains(idType)
                                && validKeyMap.get(idType).contains(item) ? 0L : 1L)
                .sum() == 0;
    }

    public static void main(String[] args) {
        assertFalse(hasValidAttributeKeys("uuid", map(new String[]{"who", "valsome"})));
        assertFalse(hasValidAttributeKeys("uuid", map(new String[]{"src", "EDITORIAL"},
                new String[]{"who", "valsome"})));
        assertTrue(hasValidAttributeKeys("uuid", map(new String[]{"src", "EDITORIAL"})));
    }

    private static Map<String, String> map(String[]... pairs) {
        Map<String, String> map = new HashMap<>();
        for (String[] pair : pairs) {
            map.put(pair[0], pair[1]);
        }
        return map;
    }
}

File: lang/src/main/scala/uk/co/zentium/langscala/Succinct1.scala
package uk.co.zentium.langscala

import org.scalatest.Matchers

object Succinct1 extends App with Matchers {

  val validKeyMap = Map(
    "logo" -> List("src"),
    "uuid" -> List("cgid", "src"),
    "pcms" -> List("cgid", "src")
  )

  def hasValidAttributeKeys(idType: String, attributes: Map[String, String]) =
    attributes.keys
      .map(item => if (validKeyMap.contains(idType) && validKeyMap(idType).contains(item)) 0 else 1)
      .sum == 0


  hasValidAttributeKeys("uuid", Map("who" -> "valsome")) should be(false)
  hasValidAttributeKeys("uuid", Map("src" -> "EDITORIAL", "who" -> "valsome")) should be(false)
  hasValidAttributeKeys("uuid", Map("src" -> "EDITORIAL")) should be(true)
}
